cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,374 @@
+namespace detail {
+
+namespace {
+
+std::unique_ptr<column> copy_with_nullable(column_view const& original,
+    bool force_mask_allocation, rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  auto copy = std::make_unique<column>(original, stream, mr);
+  if (force_mask_allocation and not original.nullable()) {
+    // Create an all-valid null mask if original has none
+    auto contents = copy->release();
+    auto bitmask = create_null_mask(original.size(), mask_state::ALL_VALID, stream, mr);
+    size_type null_count = 0;
+    copy = std::make_unique<column>(original.type(), original.size(),
+      std::move(*contents.data), std::move(bitmask), null_count, std::move(contents.children));
+    // TODO use this from PR 3172 instead
+    //auto mask = create_null_mask(target.size(), mask_state::ALL_VALID, stream, mr);
+    //copy->set_null_mask(std::move(mask), null_count);
+  }
+  return copy;
+}
+
+template <typename T>
+thrust::device_vector<T> make_gather_map(column_view const& scatter_map,
+    size_type rows, cudaStream_t stream)
+{
+  // Transform negative indices to index + rows
+  auto scatter_iter = thrust::make_transform_iterator(
+    scatter_map.begin<T>(),
+    index_converter<T>{rows});
+
+  static_assert(std::is_signed<T>::value,
+    "Need different invalid index if unsigned index types are added");
+  auto const invalid_index = static_cast<T>(-1);
+
+  // Convert scatter map to a gather map
+  auto gather_map = thrust::device_vector<T>(rows, invalid_index);
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<T>(0),
+    thrust::make_counting_iterator<T>(scatter_map.size()),
+    scatter_iter, gather_map.begin());
+
+  return gather_map;
+}
+
+template <typename MapIterator>
+void gather_bitmask(table_view const& source, MapIterator gather_map,
+    std::vector<std::unique_ptr<column>>& target, cudaStream_t stream)
+{
+  auto const device_source = table_device_view::create(source, stream);
+
+  // Make mutable table view from columns
+  auto target_views = std::vector<mutable_column_view>(target.size());
+  std::transform(target.begin(), target.end(), target_views.begin(),
+    [](auto const& col) { return static_cast<mutable_column_view>(*col); });
+  auto target_table = mutable_table_view(target_views);
+  auto device_target = mutable_table_device_view::create(target_table, stream);
+
+  // Compute block size
+  int grid_size, block_size;
+  auto bitmask_kernel = gather_bitmask_kernel<true, decltype(gather_map)>;
+  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&grid_size, &block_size, bitmask_kernel));
+
+  auto valid_counts = thrust::device_vector<size_type>(target.size());
+  bitmask_kernel<<<grid_size, block_size, 0, stream>>>(*device_source,
+    gather_map, *device_target, valid_counts.data().get());
+
+  // TODO for_each with a zip iterator?
+  auto valid_counts_host = thrust::host_vector<size_type>(valid_counts);
+  for (size_t i = 0; i < target.size(); ++i) {
```

Comment:
Generalized lambda capture (C++14). I know, not perfect...

```
for_each(target.begin(), target.end(), [i = 0, &](auto &target_col) { 
    if (target_col->nullable()) {
      target_col->set_null_count(target_col->size() - valid_counts_host[i]);
    }
    i++;
  });
```

Reply:
I went ahead and used thrust zip_iterator and host for_each